Rspec unit tests to inspec

MarkGibbons opened the pull request:
I converted the old spec tests to inspec tests.  These tests should do a pretty good job of covering the replace_or_add resource.   I based the changes to tests on the code in Dan's pull request.

I'll look at rewriting the add_to_list serverspec tests over the next month.  I needed those tests when I rewrote add_to_list a while back.

Commit:
Replace the rspec unit tests with inspec integration tests

The rspec tests were written for the heavy weight providers and did not
work when the providers were converted to custom resources. This change
deletes the old rspec tests and adds roughly equivalent inspec tests.

A chef-client handler was added for testing purposes to allow inspec tests
to check the updated vs up to date status of resources.  The line resource
has had problems with adding lines after every run.  Checking the results
after rerunning changes is important testing.

Custom inspec resources were added to make testing simpler.  Resources
were added to check the number of lines in a file and the number of
times a pattern matched the contents of a file.

damacus committed:
Merge branch 'master' into Rspec_unit_tests_to_inspec

damacus commented on Berksfile:
```diff
@@ -1,8 +1,6 @@
 
 metadata
-# cookbook 'line_test', path: 'test/cookbooks/line_test'
-# cookbook 'line_tests', path: 'spec/fixtures/cookbooks/line_tests'
```
Could you delete these lines? We no longer need them 👍

MarkGibbons replied:
Gone.

damacus commented on .kitchen.yml:
```diff
@@ -17,16 +17,27 @@ platforms:
 suites:
 - name: delete_from_list
   run_list:
+    - recipe[test::resource_status]
```
What're we trying to do with this recipe?

MarkGibbons replied:
The recipe adds a handler that reports the update status of each resource.  The tests check to see if a file was reported updated correctly by looking at the status in the report file.  Replaces the rspec idempotence checks.

damacus replied:
We could just split these tests out into their own suite. Rather than adding this on? 

Whilst it's quite nice. It's hard to get at/debug for a beginner.

MarkGibbons replied:
Removed the recipe from all of the resources except replace_or_add where it is actually used.

damacus commented on libraries/matchers.rb:
```diff
@@ -4,7 +4,7 @@
     delete_from_list: [:edit],
     delete_lines: [:edit],
-    replace_or_add: [:edit],
+    replace_or_add: [:edit]
```
The last line should have a comma on it, or was this autocorrected?

MarkGibbons replied:
I did an rspec -a by mistake.  Sorry.

MarkGibbons replied:
Upgraded to a new version of chef-dk.  There was also a cook style change that added the last comma.

damacus commented on metadata.rb:
```diff
@@ -9,6 +9,6 @@
 chef_version '>= 12.5' if respond_to?(:chef_version)
 
-%w( debian ubuntu centos redhat scientific oracle amazon ).each do |os|
+%w(debian ubuntu centos redhat scientific oracle amazon).each do |os|
```
There should be spaces inside according to cookstyle IIRC.

MarkGibbons replied:
Fixed.

damacus commented on test/fixtures/cookbooks/test/recipes/replace_or_add.rb:
```diff
@@ -22,6 +22,18 @@
 replace_or_add 'Operation 4' do
   path '/tmp/dangerfile2'
-  pattern 'ssh-rsa AAAAB3NzaC1yc2EAAAADDEADBEEF.*'
+  pattern 'ssh-dsa AAAAB3NzaC1yc2EAAAADDEADBEEF.*'
```
Why are we going from DSA --> RSA, instead of RSA to DSA?

MarkGibbons replied:
The test replaces a matching line with a nil line.  It is easier to test for and count a unique pattern.  We do lots of other tests with the ssh-rsa lines. This ssh-dsa line is here for this test only.

damacus commented on test/integration/replace_or_add/controls/replace_or_add_add_a_line_matching_pattern.rb:
```diff
@@ -0,0 +1,18 @@
+title 'replace_or_add Add a line exactly matching the pattern, pattern does not match the file'
```
You can just wrap this in a control and get rid of the dummy control file. 

It'll also make the output slightly easier to read.

MarkGibbons replied:
Wrapped in a control.

Commit:
Rspec_unit_tests_to_inspec:  Wrap tests in control blocks

Cookstyle fixes for current version of cookstyle
Add the resource_status handler only in suites where line counts are really tested

Commit:
Merge branch 'Rspec_unit_tests_to_inspec' of https://github.com/MarkGibbons/line-cookbook into Rspec_unit_tests_to_inspec

Commit:
Merge branch 'master' of https://github.com/sous-chefs/line-cookbook into Rspec_unit_tests_to_inspec

Commit:
Rspec_unit_tests_to_inspec: Dummy spec set up

Commit:
Rspec_unit_tests_to_inspec: Duplicate the test run_list in .kitchen.docken.yml
